Lombiq.Hosting.Tenants.Management/Controllers/ShellSettingsEditorController.cs

```diff
@@ -47,29 +49,56 @@ public async Task<IActionResult> Edit(ShellSettingsEditorViewModel model)
             return NotFound();
         }
 
-        var settingsDictionary = JsonConvert.DeserializeObject<Dictionary<string, string>>(model.Json);
-        var newSettings = new Dictionary<string, string>();
-
-        foreach (var key in settingsDictionary.Keys.Where(key => string.IsNullOrEmpty(settingsDictionary[key])))
+        model.Json ??= "{}";
+        if (!IsValidJson(model.Json))
         {
-            settingsDictionary[key] = null;
+            return RedirectToAction(
+                nameof(AdminController.Edit),
+                typeof(AdminController).ControllerName(),
+                new
+                {
+                    area = "OrchardCore.Tenants",
+                    id = model.TenantId,
+                });
         }
 
-        foreach (var key in settingsDictionary.Keys)
+        var settingsDictionary = new JsonConfigurationParser().ParseConfiguration(model.Json);
+        var newSettings = new Dictionary<string, string>();
+
+        var tenantSettingsPrefix = $"{model.TenantId}Prefix:";
+        var currentSettings = shellSettings.ShellConfiguration.AsEnumerable()
+            .Where(item => item.Value != null &&
+                item.Key.Contains(tenantSettingsPrefix))
+            .ToDictionary(key => key.Key.Replace(tenantSettingsPrefix, string.Empty), value => value.Value);
+
+        if (settingsDictionary?.Keys != null)
         {
-            if (shellSettings[key] != settingsDictionary[key])
+            foreach (var key in settingsDictionary.Keys)
             {
-                var tenantSettingsPrefix = $"{model.TenantId}Prefix:{key}";
-                newSettings[tenantSettingsPrefix] = settingsDictionary[key];
-                newSettings[key] = settingsDictionary[key];
+                var tenantSettingsPrefixWithKey = $"{tenantSettingsPrefix}{key}";
+                if (shellSettings[key] != settingsDictionary[key])
+                {
+                    newSettings[tenantSettingsPrefixWithKey] = settingsDictionary[key];
+                    newSettings[key] = settingsDictionary[key];
+                }
             }
         }
 
-        // Try to acquire a lock before using the scope, so that a next process gets the last committed data.
-        var (locker, locked) = await _distributedLock.TryAcquireLockAsync("SHELL_SETTINGS_EDITOR_LOCK", TimeSpan.MaxValue);
+        var deletableKeys = currentSettings
+            .Where(item => settingsDictionary == null || !settingsDictionary.ContainsKey(item.Key))
+            .Select(item => item.Key);
+
+        foreach (var key in deletableKeys)
+        {
+            var tenantSettingsPrefixWithKey = $"{tenantSettingsPrefix}{key}";
+            newSettings[key] = null;
+            newSettings[tenantSettingsPrefixWithKey] = null;
+        }
+
+        var (locker, locked) = await _distributedLock.TryAcquireLockAsync("SHELL_SETTINGS_EDITOR_LOCK", TimeSpan.FromSeconds(10));
```

Comment:
Since this is a global identifier:

```suggestion
        var (locker, locked) = await _distributedLock.TryAcquireLockAsync("LOMBIQ_HOSTING_TENANTS_MANAGEMENT_SHELL_SETTINGS_EDITOR_LOCK", TimeSpan.FromSeconds(10));
```